File: examples/F90/grib_copy_message.f90
! Copyright 2005-2016 ECMWF.
!
! This software is licensed under the terms of the Apache Licence Version 2.0
! which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
! 
! In applying this licence, ECMWF does not waive the privileges and immunities granted to it by
! virtue of its status as an intergovernmental organisation nor does it submit to any jurisdiction.
!
!
!
!  Description: how to copy a GRIB message in memory
!
!
!
program copy
  use eccodes
  implicit none
  integer                            :: err, centre
  integer(kind=kindOfSize)           :: byte_size
  integer                            :: infile,outfile
  integer                            :: igrib_in
  integer                            :: igrib_out
  character(len=1), dimension(:), allocatable :: message
  character(len=32)                  :: product_kind


  call codes_open_file(infile,'../../data/constant_field.grib1','r')
  call codes_open_file(outfile,'out.copy.grib1','w')

  ! A new GRIB message is loaded from file
  ! igrib_in is the GRIB id to be used in subsequent calls
  call codes_grib_new_from_file(infile, igrib_in)

  call codes_get_message_size(igrib_in, byte_size)
  allocate(message(byte_size), stat=err)

  call codes_copy_message(igrib_in, message)

  call codes_new_from_message(igrib_out, message)

  call codes_get(igrib_out, 'kindOfProduct', product_kind)
  write(*,*) 'kindOfProduct=',product_kind

  centre=80
  call codes_set(igrib_out, 'centre', centre)

  ! Write message to a file
  call codes_write(igrib_out, outfile)

  call codes_release(igrib_out)
  call codes_release(igrib_in)
  call codes_close_file(infile)
  call codes_close_file(outfile)
  deallocate(message)

end program copy

File: tests/grib_copy_message.cc
/*
 * (C) Copyright 2005- ECMWF.
 *
 * This software is licensed under the terms of the Apache Licence Version 2.0
 * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * In applying this licence, ECMWF does not waive the privileges and immunities granted to it by
 * virtue of its status as an intergovernmental organisation nor does it submit to any jurisdiction.
 */
#include <stdio.h>
#undef NDEBUG
#include <assert.h>
#include "eccodes.h"

int main(int argc, char* argv[])
{
    FILE* in                     = NULL;
    codes_handle* source_handle  = NULL;
    int err                      = 0;
    size_t totalLength = 0, size = 0;
    unsigned char* buffer        = NULL;
    codes_handle* new_handle     = NULL;

    assert (argc == 3);

    in = fopen(argv[1], "rb");
    assert(in);

    source_handle = codes_handle_new_from_file(0, in, PRODUCT_GRIB, &err);
    assert(source_handle);

    // How big is the input GRIB message?
    CODES_CHECK(codes_get_message_size(source_handle, &totalLength), 0);

    // Allocate a buffer large enough to hold the message
    buffer = (unsigned char*)malloc(totalLength);
    size = totalLength;

    // Take a copy of the message into buffer. Now we own it
    CODES_CHECK(codes_get_message_copy(source_handle, buffer, &size), 0);
    assert(size == totalLength);
    codes_handle_delete(source_handle);

    // Now buffer contains a copy of the message
    new_handle = codes_handle_new_from_message(0, buffer, totalLength);
    assert(new_handle);

    // Change something and write it out
    CODES_CHECK(codes_set_long(new_handle, "hour", 18), 0);
    CODES_CHECK(codes_write_message(new_handle, argv[2], "w"), 0);
    codes_handle_delete(new_handle);
    free(buffer);

    fclose(in);

    return 0;
}
